File: Python/Sorting/quick_Sort.py
# QUICK SORT -> Divide and Conquer Algorithm

# Best: O(nlog(n)) Time Complexity | O(nlog(n)) Space Complexity
# Average: O(nlog(n)) Time Complexity | O(nlog(n)) Space Complexity
# Worst: O(n^2) Time Complexity | O(nlog(n)) Space Complexity

def quick_sort(array):
    """Function to hold the main implementation of quick sort algorithm"""
    quick_sort_aux(array, 0, len(array) - 1)
    return array

def quick_sort_aux(array, start_idx, end_idx):
    """Function to sort array"""
    if start_idx >= end_idx:
        # Array sorted
        return
    pivot_idx = start_idx
    left_idx = start_idx + 1
    right_idx = end_idx
    while right_idx >= left_idx:
        # Main loop to sort array
        if array[left_idx] > array[pivot_idx] and array[right_idx] < array[pivot_idx]:
            swap(left_idx, right_idx, array)
        if array[left_idx] <= array[pivot_idx]:
            left_idx += 1
        if array[right_idx] >= array[pivot_idx]:
            right_idx -= 1
    swap(pivot_idx, right_idx, array)
    left_sub_array_smaller = right_idx - 1 - start_idx < end_idx - (right_idx - 1) # If left sub array is smaller than right sub array
    if left_sub_array_smaller:
        # Start sorting the smaller sub array first -> Best approach regarding space complexity (so log of n is not exceede)
        quick_sort_aux(array, start_idx, right_idx - 1)
        quick_sort_aux(array, right_idx + 1, end_idx)
    else:
        quick_sort_aux(array, right_idx + 1, end_idx)
        quick_sort_aux(array, start_idx, right_idx - 1)

def swap(i, j, array):
    """Function to swap two elements in the same array"""
    array[i], array[j] = array[j], array[i]

# Main program
print(quick_sort([1, -12, 9, 54, -11, 75, 8, 14]))
    

File: C/Sorting/quick_Sort.c
#include<stdio.h>

#define N 8

//Function Signature
void quickSort(int*, int, int); //Function to sort the array recursively
int quickSortAux(int*,int, int); //Function to partition the array
void swap(int*, int*); //Function to swap two elements

int main(){
    int myArray[N] = {1, -12, 9, 54, -11, 75, 8, 14};
    int i;
    int arrayLen = N;
    
    quickSort(myArray, 0, arrayLen - 1);
    for(i = 0; i < N; i++){
        printf(" %d ", myArray[i]);
    }
}

void quickSort(int *myArray, int startIdx, int endIdx){
    int partIdx;

    if(startIdx < endIdx){
        partIdx = quickSortAux(myArray, startIdx, endIdx); //Partion index, now at correct place
        quickSort(myArray, startIdx, partIdx - 1); //Sort array before partIdx
        quickSort(myArray, partIdx + 1, endIdx); //Sort array after partIdx
    }
}

int quickSortAux(int *myArray, int startIdx, int endIdx){
    int pivot = myArray[endIdx]; //Pivor is the is placed at the last elemtn
    int i = startIdx - 1;
    int j;

    for(j = startIdx; j <= endIdx - 1; j++){
        //All the smaller elements than pivot to the left
        //All the greater elements than pivot to the right
        if(myArray[j] < pivot){
            //If current element is samller than pivot
            i++; //Increment index
            swap(&myArray[i], &myArray[j]);
        }
    }
    swap(&myArray[i + 1], &myArray[endIdx]);
    return (i + 1); //Partion index 
}

void swap(int *i, int *j){
    int temp = 0; 

    temp = *i;
    *i = *j;
    *j = temp;
}